session/direct-ipc.c

```diff
@@ -0,0 +1,241 @@
+#define _POSIX_C_SOURCE 200809L
+#include <errno.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/sysmacros.h>
+#include <sys/wait.h>
+#include <xf86drm.h>
+#include <linux/major.h>
+#include <wlr/util/log.h>
+#include "session/direct-ipc.h"
+
+enum { DRM_MAJOR = 226 };
+
+#ifdef HAS_LIBCAP
+#include <sys/capability.h>
+
+static bool have_permissions(void) {
+	cap_t cap = cap_get_proc();
+	cap_flag_value_t val;
+
+	if (!cap || cap_get_flag(cap, CAP_SYS_ADMIN, CAP_PERMITTED, &val) || val != CAP_SET) {
+		wlr_log(L_ERROR, "Do not have CAP_SYS_ADMIN; cannot become DRM master");
+		cap_free(cap);
+		return false;
+	}
+
+	cap_free(cap);
+	return true;
+}
+#else
+static bool have_permissions(void) {
+	if (geteuid() != 0) {
+		wlr_log(L_ERROR, "Do not have root privileges; cannot become DRM master");
+		return false;
+	}
+
+	return true;
+}
+#endif
+
+static void send_msg(int sock, int fd, void *buf, size_t buf_len) {
+	char control[CMSG_SPACE(sizeof(fd))] = {0};
+	struct iovec iovec = { .iov_base = buf, .iov_len = buf_len };
+	struct msghdr msghdr = {0};
+
+	if (buf) {
+		msghdr.msg_iov = &iovec;
+		msghdr.msg_iovlen = 1;
+	}
+
+	if (fd >= 0) {
+		msghdr.msg_control = &control;
+		msghdr.msg_controllen = sizeof(control);
+
+		struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msghdr);
+		*cmsg = (struct cmsghdr) {
+			.cmsg_level = SOL_SOCKET,
+			.cmsg_type = SCM_RIGHTS,
+			.cmsg_len = CMSG_LEN(sizeof(fd)),
+		};
+		*(int *)CMSG_DATA(cmsg) = fd;
+	}
+
+	ssize_t ret;
+	do {
+		ret = sendmsg(sock, &msghdr, 0);
+	} while (ret < 0 && errno == EINTR);
+}
+
+static ssize_t recv_msg(int sock, int *fd_out, void *buf, size_t buf_len) {
+	char control[CMSG_SPACE(sizeof(*fd_out))] = {0};
+	struct iovec iovec = { .iov_base = buf, .iov_len = buf_len };
+	struct msghdr msghdr = {0};
+
+	if (buf) {
+		msghdr.msg_iov = &iovec;
+		msghdr.msg_iovlen = 1;
+	}
+
+	if (fd_out) {
+		msghdr.msg_control = &control;
+		msghdr.msg_controllen = sizeof(control);
+	}
+
+	ssize_t ret;
+	do {
+		ret = recvmsg(sock, &msghdr, MSG_CMSG_CLOEXEC);
+	} while (ret < 0 && errno == EINTR);
+
+	if (fd_out) {
+		struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msghdr);
+		*fd_out = cmsg ? *(int *)CMSG_DATA(cmsg) : -1;
+	}
+
+	return ret;
+}
+
+enum msg_type {
+	MSG_OPEN,
+	MSG_SETMASTER,
+	MSG_DROPMASTER,
+	MSG_END,
+};
+
+struct msg {
+	enum msg_type type;
+	char path[256];
+};
+
+static void communicate(int sock) {
+	struct msg msg;
+	int drm_fd = -1;
+	bool running = true;
+
+	while (running && recv_msg(sock, NULL, &msg, sizeof(msg)) >= 0) {
+		switch (msg.type) {
+		case MSG_OPEN:
+			errno = 0;
+
+			// These are the same flags that logind opens files with
+			int fd = open(msg.path, O_RDWR|O_CLOEXEC|O_NOCTTY|O_NONBLOCK);
```

Comment:
We probably want to have a whitelist of files that you can open so that a rogue child process can't have free reign over the system.